src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationPartFactory.cs

```diff
@@ -27,5 +34,52 @@ public abstract class ApplicationPartFactory
         /// The context name. By default, value of this parameter is <see cref="DefaultContextName"/>.
         /// </param>
         public abstract IEnumerable<ApplicationPart> GetApplicationParts(Assembly assembly, string context);
+
+        /// <summary>
+        /// Gets the <see cref="ApplicationPartFactory"/> for the specified assembly.
+        /// <para>
+        /// An assembly may specify an <see cref="ApplicationPartFactory"/> using <see cref="ProvideApplicationPartFactoryAttribute"/>.
+        /// Otherwise, <see cref="ApplicationPartFactory.Default"/> is used.
+        /// </para>
+        /// </summary>
+        /// <param name="assembly">The <see cref="Assembly"/>.</param>
+        /// <returns>An instance of <see cref="ApplicationPartFactory"/>.</returns>
+        public static ApplicationPartFactory GetApplicationPartFactory(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            var provideAttribute = assembly.GetCustomAttribute<ProvideApplicationPartFactoryAttribute>();
```

Comment:
nit: I would just name the attribute ApplicationPartFactoryAttribute

Reply:
We landed on the attribute name as part of the wireframe design. Unless we have a very good reason to rename it, I'm going to leave it as is.